TREQ-21: Throw FileNotFoundException when resource is not found

SourcePond opened the pull request:
- Check on construction time whether the URL points to a local resource
- Always throw a FileNotFoundException when the URL points to a local resource and url.openConnection throws an IOException
- Added OSGi tests

SourcePond commented:
This pull-request replaces https://github.com/apache/tiles-request/pull/1

michaelsembwever commented on tiles-request-api/pom.xml:
```diff
@@ -31,10 +31,21 @@ under the License.
+        <url.version>2.5.2</url.version>
+        <felix.version>5.6.4</felix.version>
+        <slf4j-api.version>1.7.25</slf4j-api.version>
+        <logback.version>1.2.2</logback.version>
```
Trivial:
 We don't specify dependency versions by maven properties in the tiles projects.
 Please put these in with the dependency elements.

michaelsembwever commented on tiles-request-api/src/main/java/org/apache/tiles/request/locale/URLApplicationResource.java:
```diff
@@ -21,20 +21,18 @@
+import java.net.*;
+import java.util.HashSet;
+import java.util.Locale;
+import java.util.Set;
```
Trivial:
 the import order was correct how it was. the project puts java.* imports first, and doesn't wildcard imports.

michaelsembwever commented on tiles-request-api/src/main/java/org/apache/tiles/request/locale/URLApplicationResource.java:
```diff
@@ -44,10 +42,45 @@
+     * {@link IOException} will be thrown directly. If a url has a local protocol, then any {@link IOException}
+     * will be caught and transformed into a {@link FileNotFoundException}.
+     */
+    static final String REMOTE_PROTOCOLS_PROPERTY = "tiles.remoteProtocols";
```
Trivial:
 according to Java's declaration order, this should appear before `LOG` is declared.

michaelsembwever replied:
+1

michaelsembwever commented on tiles-request-api/src/main/java/org/apache/tiles/request/locale/URLApplicationResource.java:
```diff
@@ -44,10 +42,45 @@
+        String protocolsProp = getProperty(REMOTE_PROTOCOLS_PROPERTY);
+        if (protocolsProp != null) {
+            for (String protocol : protocolsProp.split(";")) {
+                remoteProtocols.add(protocol.trim());
```
this is adding a feature to the code. it will need a bit more documentation, and be a bit clearer in the issue and commit message, as currently it's titled as a bug fix.

michaelsembwever replied:
perfect. thanks @SourcePond

michaelsembwever commented on tiles-request-api/src/main/java/org/apache/tiles/request/locale/URLApplicationResource.java:
```diff
@@ -44,10 +42,45 @@
     /** if the URL matches a file, this is the file. */
     private File file;
+    /** if the URL points to a local resource */
+    private boolean local;
```
i think we can make these fields `final`?

michaelsembwever reviewed (changes requested):
Thanks @SourcePond, the PR looks awesome.

I've added some trivial comments regarding the project's code style and so on…

Otherwise I need some time to check out the tests, as you're added new testing framework/libraries…

My other question was that you were making remote url protocols pluggable, at least in how we detect which protocols are local or not…
What else is required to add a protocol to work with `url.openConnecton()`?

I'm thinking that if REMOTE_PROTOCOLS can be extended by a system property, then it should be documented, including an example use-case or what else needs to be done for additional protocols.

SourcePond commented:
I added some documentation which should also explain how an additional remote protocol can be added to the system. Additionally, I enhanced the task a bit so it should explain the problem in more detail.

michaelsembwever commented:
> I added some documentation which should also explain how an additional remote protocol can be added to the system. Additionally, I enhanced the task a bit so it should explain the problem in more detail.

Thanks @SourcePond. Very much appreciated! Especially the response time :-)

michaelsembwever commented on tiles-request-api/src/test/java/org/apache/tiles/request/locale/URLApplicationResourceTest.java:
```diff
@@ -20,20 +20,24 @@
-import java.net.URISyntaxException;
-import java.net.URL;
+import java.lang.reflect.Field;
+import java.net.*;
```
trivial: no wildcards please.

SourcePond replied:
I'm sorry, I adjusted my IDE settings and it should not happen again now ;-)

michaelsembwever commented on tiles-request-api/src/test/java/org/apache/tiles/request/locale/URLApplicationResourceTest.java:
```diff
@@ -201,4 +224,36 @@ public void testGetInputStreamWithSpace() throws IOException {
+        try {
+            resource.getInputStream();
+        } catch (FileNotFoundException e) {
+            // expected
```
if this is expected, then shouldn't the method be annotated `@Test(expected=FileNotFoundException.class)` ?

michaelsembwever commented on tiles-request-api/src/test/java/org/apache/tiles/request/locale/URLApplicationResourceTest.java:
```diff
@@ -201,4 +224,36 @@ public void testGetInputStreamWithSpace() throws IOException {
+        } catch (FileNotFoundException e) {
+            fail("FileNotFoundException not allowed here");
+        } catch (IOException e) {
+            assertEquals(EXPECTED_MESSAGE, e.getMessage());
```
same comment here. shouldn't we also rethrow, and make it `@Test(expected=IOException.class)`

SourcePond replied:
In my opinion, we should catch the IOException here, and, should assert that it was actually thrown by the TestUrlStreamHandler. This should make sure that the test does not accidentally pass even when an unexpected IOException occurs.

michaelsembwever commented on tiles-request-api/src/test/java/org/apache/tiles/request/osgi/URLBundleApplicationResourceTest.java:
```diff
@@ -0,0 +1,83 @@
+            resource.getLastModified();
+            fail("Exception expected");
+        } catch (FileNotFoundException e) {
+            // expected
```
same comment about expected exceptions.

michaelsembwever commented on tiles-request-api/src/test/java/org/apache/tiles/request/osgi/URLBundleApplicationResourceTest.java:
```diff
@@ -0,0 +1,83 @@
+            resource.getInputStream();
+            fail("Exception expected");
+        } catch (FileNotFoundException e) {
+            // expected
```
same comment about expected exceptions.

michaelsembwever commented:
i've finished my review and apart from the comments about `@Test(expected=…)` this is :+1:  from me.

michaelsembwever commented:
@SourcePond  i'm happy to merge this. thanks for the contribution!

Before I merge it, would you mind squashing the commits to one commit, writing up a summarised message (that includes a reference to this PR)?

SourcePond committed:
TREQ-21: Convert IOException to FileNotFoundException on local protocols
- Separate between local and remote URL protocols, check on construction time whether the URL points to a local resource
- Always throw a FileNotFoundException when the URL points to a local resource and url.openConnection throws an IOException
- Added system property 'tiles.remoteProtocols' which can be used to specify additional remote protocols
- Added OSGi tests including PAX-EXAM dependencies
- PR: https://github.com/apache/tiles-request/pull/2

SourcePond commented:
I squashed all changes into 1 commit, I'm glad that I could help

michaelsembwever commented:
Has been committed upstream.
